Partially implemented FontDescription

gsingh93 opened the pull request:
Implements the majority of FontDescription and its methods.

GuillaumeGomez commented:
Nice one ! Thanks !

gsingh93 commented:
Should I be using any `glib::translate` methods for the conversions? And should the wrapped `pointer` be public, or is there some glib trait I should implement to expose it?

GuillaumeGomez commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,117 @@
+    pub fn copy(desc: &FontDescription) -> Option<FontDescription> {
+        let tmp = unsafe { ffi::pango_font_description_copy(desc.pointer) };
+
+        if tmp.is_null() {
```
Take a look here : https://github.com/rust-gnome/gdk/blob/master/src/pixbuf/loader.rs#L29

GuillaumeGomez commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    pub fn set_family(desc: &mut FontDescription, family: &str) {
+        let c_family = CString::new(family).unwrap();
```
Take a look here : https://github.com/rust-gnome/gdk/blob/61632df14152b4a845fccdd77930c2d9941a50dc/src/rt.rs#L35

GuillaumeGomez commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,117 @@
+        unsafe { ffi::pango_font_description_set_family_static(desc.pointer, c_family.as_ptr()) };
+    }
+
+    pub fn get_family(desc: &FontDescription) -> Option<String> {
```
Take a look here : https://github.com/rust-gnome/gdk/blob/61632df14152b4a845fccdd77930c2d9941a50dc/src/rt.rs#L25

gkoz commented:
Using the translations would be preferable. Do you need any help with that? I don't think I've had any feedback about using those APIs yet.
That way publishing the pointer isn't needed.

gkoz commented:
Oh, as a non-`GObject` type the best convention for wrapping this might not be settled upon.

gkoz commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,117 @@
+
+    pub fn set_family_static(desc: &mut FontDescription, family: &'static str) {
+        let c_family = CString::new(family).unwrap();
+        unsafe { ffi::pango_font_description_set_family_static(desc.pointer, c_family.as_ptr()) };
```
The contract is that the pointer stays alive for the lifetime of the object but it's freed at the end of the block.
I don't believe this optimization be worth the hassle so let's just drop this function.

gsingh93 commented:
So I used the translate functions for the strings (can you double check I used the write methods for the write ownerships?), but I'm not sure about using them for the object itself. FontDescription isn't a GObject, it's a GBoxed. Does that make a difference?

gkoz commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,115 @@
+
+    pub fn set_family_static(desc: &mut FontDescription, family: &'static str) {
+        unsafe {
+            ffi::pango_font_description_set_family_static(desc.pointer, family.to_glib_none().0)
```
(Repeating so it doesn't get lost accidentally)

The contract is that the pointer stays alive for the lifetime of the object but it's freed at the end of the block.
I don't believe this optimization be worth the hassle so let's just drop this function.

gsingh93 replied:
I don't mind dropping it, but at the same time I don't see the problem with leaving it.

gkoz replied:
The lifetime of the pointer is the enclosing block. The required lifetime is the lifetime of the FontDescription object. Effectively this always is a dangling pointer.

gsingh93 replied:
Oh, right. I'll remove this.

gsingh93 replied:
We could also `mem::forget` the `CString`. Do you still want me to remove it?

gkoz commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,115 @@
+    pub fn to_string(desc: &FontDescription) -> String {
+        unsafe {
+            let tmp = ffi::pango_font_description_to_string(desc.pointer);
+            debug_assert!(!tmp.is_null());
```
The NULL check is handled by `from_glib_full` in case the return type is `String` and not `Option<String>`.

gkoz commented:
> So I used the translate functions for the strings (can you double check I used the right methods for the right ownerships?)

`set_family` seems okay. There's also `from_string`.

--

Methods on a struct take the receiver as `&self` not as `smth: &SomeType`.

gkoz commented on src/widgets/font_description.rs:
```diff
@@ -0,0 +1,115 @@
+use ffi;
+use enums;
+use glib::translate::{ToGlibPtr, from_glib_full, from_glib_none};
+use glib::to_bool;
```
I might kill `to_bool` someday so I recommend using `from_glib` instead ;)

gsingh93 commented:
I made the suggested fixes and also exported `FontDescription` at the top level. I'm interpreting your comments about GObject/GBoxed conversion to mean "don't worry about it for now".

gkoz commented:
The fixes look good, except where you had `&mut FontDescription` it should stay `&mut self`. Contrary to `GObject` descendants we actually own the pointer so mutable method here make perfect sense.

> I'm interpreting your comments about GObject/GBoxed conversion to mean "don't worry about it for now".

Not quite. Rather, we should use `glib::translate` and look out for any unforeseen issues. (One slight issue is that the translations don't particularly care about the difference between `*const` and `*mut` but it's not a stopper). `to_glib_full` would have to just be `{ unimplemented!() }`, because there's no reference counting for this type.

`from_string` only returns `NULL` if we give it `NULL` as an argument, so `Option` isn't needed there.

gkoz commented on src/pango.rs:
```diff
@@ -16,7 +17,8 @@ pub use self::widgets::{
     Matrix,
-    GlyphString
+    GlyphString,
+    FontDescription
```
Rust style recommends trailing commas in such cases to reduce churn in the future (note how the above line had to change because it lacked the comma).

gsingh93 commented:
> Contrary to GObject descendants we actually own the pointer so mutable method here make perfect sense.

I don't understand why GObject methods don't need to use `&mut self`. Who owns the pointer?

> Not quite. Rather, we should use glib::translate and look out for any unforeseen issues. (One slight issue is that the translations don't particularly care about the difference between *const and *mut but it's not a stopper). to_glib_full would have to just be { unimplemented!() }, because there's no reference counting for this type.

Can you hop on to IRC so we can talk about this?

Regarding returning options from constructors, I noticed most of the types in `gtk` return `Option`s from their constructors, even though I don't see anything in the documentation about returning `NULL` (one example is `Label`). Why is this done?

gkoz commented:
> I don't understand why GObject methods don't need to use &mut self. Who owns the pointer?

`GObject`s are reference counted so they are very similar to `Rc<RefCell<T>>`, there isn't a single owner and a unique borrow of our wrapper struct doesn't mean anything.

> Regarding returning options from constructors, I noticed most of the types in gtk return Options from their constructors.

That approach was a mistake and it's being gradually rooted out. The constructors aren't supposed to ever return `NULL`.

gsingh93 commented:
I added another commit with the requested changes, as well as a partial `FromGlibPtr`/`ToGlibPtr` implementation. I can rebase into one commit before merging.

1. Is there a way for me to implement `from_glib_full` without copying? It's not a good idea to leave the previous owner with access to the structure, so the free is necessary, right?
2. How can I implement `to_glib_none`? What would the storage type be?
3. Should I make `FontDescription` a newtype (i.e. a tuple struct)?

gkoz commented:
> Is there a way for me to implement from_glib_full without copying?

This essentially is a move. You don't have to do anything except the `NULL` check, the ownership (and the responsibility to `unref` or `free`) is transferred to you, that's it. The example linked below illustrates that.

> How can I implement to_glib_none? What would the storage type be?

Same as with [`FrameTimings` from `gdk`](https://github.com/rust-gnome/gdk/blob/master/src/frame_timings.rs#L63-69). The trait is implemented for a reference and that reference is put into storage guaranteeing that the pointer is valid as long as the `Stash` and therefore the original reference is alive. This might seem an overkill when you're just dealing with a single pointer but allows robust support for arrays and nested structures.

> Should I make FontDescription a newtype (i.e. a tuple struct)?

I'd prefer this yes. Hopefully, a tuple struct is not considered equivalent to a newtype. It seems to be a popular and reasonably concise way to wrap pointers.

gsingh93 commented:
> Hopefully, a tuple struct is not considered equivalent to a newtype.

I was using the terminology from [the book](https://doc.rust-lang.org/book/tuple-structs.html)

I believe I've addressed all of the issues, thanks for the helpful comments!

gkoz commented:
Everything seems good.

Implementing `ToString` is a nice idea. Considering that there is
```rust
impl<T> ToString for T where T: Display + ?Sized
```
it may be even better to implement `Display` instead.

> I was using the terminology from the book

You're right, I guess my understanding of the newtype term was too limited.

GuillaumeGomez commented:
When everything's good, please notify for merging.

gsingh93 commented:
Adding Display is a good idea, updated.

gkoz commented:
:+1: 
@GuillaumeGomez

GuillaumeGomez commented:
@gkoz: Thanks for helping him ! =D
@gsingh93: Thanks for your work !

gkoz commented:
There's one more thing I forgot: I'd like `From/ToGlibPtr` impls to have `#[inline]` attribute.
